### PR TITLE
feat(fallback-pricefeed): Create FallBackPriceFeed and ForexPriceFeed

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

The TraderMadePriceFeed has been empirically unstable, inspiring this new pricefeed which allows the user to specify a priority list of pricefeeds to "fallback" through. For example, the user might use this new pricefeed to first try to get a price from the TraderMadePriceFeed for the EURUSD identifier, but if that fails then they can use the more reliable but less granular ForexPriceFeed for daily EURUSD prices.

**Summary**

**Details**

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [X]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js [hunk 1]
```diff
@@ -436,11 +484,22 @@ async function createPriceFeed(logger, web3, networker, getTime, config) {
...
   async function _createMedianizerPriceFeed(medianizerConfig) {
-    const priceFeedsToMedianize = [];
-    for (const _priceFeedConfig of medianizerConfig.medianizedFeeds) {
-      // The medianized feeds should inherit config options from the parent config if it doesn't define those values
+    const priceFeedsToMedianize = await _createConstituentPriceFeed(medianizerConfig.medianizedFeeds);
+    if (!priceFeedsToMedianize) return null;
+    return new MedianizerPriceFeed(priceFeedsToMedianize, medianizerConfig.computeMean);
+  }
+
+  async function _createFallBackPriceFeed(fallbackConfig) {
```
Comment (nicholaspai) [inline packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js:493]: Refactored this internal method a bit since we often want to create multiple pricefeeds from an array of configs

File: packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js [hunk 2]
```diff
@@ -436,11 +484,22 @@ async function createPriceFeed(logger, web3, networker, getTime, config) {
...
+    const orderedPriceFeeds = await _createConstituentPriceFeed(fallbackConfig.orderedFeeds);
+    if (!orderedPriceFeeds) return null;
+    return new FallBackPriceFeed(orderedPriceFeeds);
+  }
+
+  async function _createConstituentPriceFeed(priceFeedConfigs) {
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js:499]: Would plural make sense considering it takes in and returns an array?

```suggestion
  async function _createConstituentPriceFeeds(priceFeedConfigs) {
```

File: packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js [hunk 3]
```diff
@@ -399,19 +399,41 @@ const defaultConfigs = {
...
-    hourlyLookback: 604800,
-    ohlcPeriod: 10 // CNYUSD only available at 10 minute granularity
+    type: "fallback",
+    orderedFeeds: [
+      {
+        type: "tradermade",
+        pair: "CNYUSD",
+        minTimeBetweenUpdates: 600,
+        minuteLookback: 7200,
+        hourlyLookback: 604800,
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js:409]: Not necessarily suggesting we change it, but why are the default hourly lookbacks for these feeds 7 days? I could see making it 3 hours so we ensure that it at least covers the full 2 hours of the minutely lookback, but it seems like there's no reason the bots would need 7 days? Reducing this may also help with reliability since we're not asking for as much data.

File: packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js [hunk 4]
```diff
@@ -399,19 +399,41 @@ const defaultConfigs = {
...
+        ohlcPeriod: 10 // CNYUSD only available at 10 minute granularity
+      },
+      {
+        type: "forexdaily",
+        base: "CNY",
+        symbol: "USD",
+        lookback: 604800
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js:416]: Similar question as above -- could make sense to shorten this to 1 or 2 days (if the feed is daily). We basically want to ensure we _at least_ cover the standard 2 hour period, but any more than that seems excessive. But maybe I'm missing something?

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    // If no pricefeeds return successfully, indicate failure.
+    return null;
+  }
+
+  async getHistoricalPrice(time, verbose = false) {
+    // If failure to fetch any constituent historical prices, then throw
+    // array of errors.
+    let errors = [];
+    for (let _priceFeed of this.priceFeeds) {
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js:38]: nice and simple -- I like it.

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      }
+    }
+
+    throw errors;
+  }
+
+  // Note: This method will fail if one of the pricefeeds has not implemented `getHistoricalPricePeriods`, which
+  // is basically every price feed except for the CryptoWatchPriceFeed.
+  getHistoricalPricePeriods() {
+    throw new Error("getHistoricalPricePeriods Unimplemented for FallBackPriceFeed");
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js:53]: +1 for implementing and erroring.

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  // Gets the update time for the first pricefeed that successfully updated.
+  getLastUpdateTime() {
+    for (let _priceFeed of this.priceFeeds) {
+      const lastUpdateTime = _priceFeed.getLastUpdateTime();
+      if (!lastUpdateTime) continue;
+      return lastUpdateTime;
+    }
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js:62]: Falling back on methods like this are a little weird. I'm not exactly sure how to handle it.

What about taking the latest of all successful return values? Not strongly opinionated, but this method seems to rely on the update having _never_ succeeded, which may not be true...? Really not sure :/

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    for (let _priceFeed of this.priceFeeds) {
+      try {
+        return await _priceFeed.update();
+      } catch (err) {
+        errors.push(err);
+        continue;
+      }
+    }
+
+    throw errors;
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js:93]: What happens if the historical price query _succeeds_, but returns empty data on update. Then when you call the method to grab the historical prices, it finds no data and falls back. However, since we never called update on the fallback, it has no data either. Thoughts? Should we update _all_ fallbacks just in case? What do you think?

This is another reason that it might be nice to ultimately drop the update method entirely.

File: packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+    // 2. Send request.
+    const historyResponse = await this.networker.getJson(url);
+
+    // 3. Check responses.
+    if (
+      !historyResponse ||
+      !historyResponse.rates ||
+      Object.keys(historyResponse.rates).length === 0 ||
+      !historyResponse.rates[Object.keys(historyResponse.rates)[0]][this.symbol]
+    ) {
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js:191]: Thoughts on doing this to check every element and simplify a bit?

```suggestion
    if (
      !historyResponse?.rates
      Object.keys(historyResponse.rates).length === 0 ||
      Object.values(historyResponse.rates).some(rate => !rate[this.symbol])
    ) {
```

**Review by mrice32 — COMMENTED** [review]
Looks awesome! Mostly minor comments around the fallback price feed and other minutia. Thanks for putting this together so quickly!!

File: packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js (hunk 3 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js:409]: Well we want this to at least cover the period of time in which no data is being served over the weekend, so if the `minuteLookback` starts failing (i.e. returns NO data) 2 hours into the weekend (which begins at 22:00 UTC on Friday), then we would want `hourLookback` to cover a bit less than 2 days (but probably more over extended holiday periods). So I think a worst case setting here is 4 days, but I agree 7 days is too much. WDYT

File: packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js (hunk 4 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js:416]: That's reasonable, I'm going to set to 3 days for the reason mentioned before

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js (hunk 7 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js:62]: Shouldn't this return the update time of the pricefeed that is used for the pricefeed, i.e. the first pricefeed to return successfully?

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js (hunk 8 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js:93]: > What happens if the historical price query succeeds, but returns empty data on update
Hmmmm I think the issue you're pointing out is that  its possible right now that `update` succeeds but `historicalPrice` throws an error. So from that perspective it would make sense to `update` them all.

The downside is unneccessary updates and fetched data, which makes this pricefeed a bit less powerful.

And yes, the `update` is definitely overkill here, as it would be much better if `await getHistoricalPrice` was all we needed to call get a historical price.

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js (hunk 7 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js:62]: Ok so as mentioned [here](https://github.com/UMAprotocol/protocol/pull/2718#discussion_r595626631), I'm leaning towards updating ALL pricefeeds (and catching + ignoring any errors), and then using the latest update time as the update time here.

**Comment by nicholaspai** [comment]
@mrice32 responded to your comments in [this commit](https://github.com/UMAprotocol/protocol/pull/2718/commits/99bb14f4da0d3ea71a7f5cfc263fda7b2296c8b1) (with comments)

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js (hunk 7 above)
Comment (mrice32) [reply packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js]: Yeah, I think that makes sense.

> Shouldn't this return the update time of the pricefeed that is used for the pricefeed, i.e. the first pricefeed to return successfully?
Ideally, yes, but I don't think it's true that this method errors when getCurrentPrice or getHistoricalPrice error, so you would effectively always get the first feed if that feed had ever succeeded.

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js (hunk 8 above)
Comment (mrice32) [reply packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js]: Agreed. It does make this a bit less powerful.

File: packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js (hunk 3 above)
Comment (mrice32) [reply packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js]: Seems reasonable to me!

File: packages/financial-templates-lib/test/price-feed/CreatePriceFeed.js [hunk 10]
```diff
@@ -760,6 +764,98 @@ contract("CreatePriceFeed.js", function(accounts) {
...
+        }
+      ]
+    };
+
+    const validFallbackFeed = await createPriceFeed(logger, web3, networker, getTime, config);
+
+    assert.equal(validFallbackFeed.priceFeeds[0].lookback, lookbackOverride);
+  });
+
+  it("Fallback feed cannot have 0 nested feeds to medianize", async function() {
```
Comment (mrice32) [inline packages/financial-templates-lib/test/price-feed/CreatePriceFeed.js:819]:
```suggestion
  it("Fallback feed cannot have 0 nested feeds", async function() {
```

File: packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+    this.historicalPricePeriods = newHistoricalPricePeriods;
+    this.lastUpdateTime = currentTime;
+  }
+
+  _secondToDateTime(inputSecond) {
+    return moment.unix(inputSecond).format("YYYY-MM-DD");
+  }
+  _dateTimeToSecond(inputDateTime) {
+    return moment(inputDateTime, "YYYY-MM-DD").unix();
+  }
```
Comment (mrice32) [inline packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js:228]: I'm not sure how precise we want to be around these seconds -> day conversions, but the ECB data is [published at 16:00 CET](https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html) (computed at 14:15 CET, which might also be a good choice for stamping it), which is 3:00PM UTC and 11:AM EST. Using the terminology from this price feed, I would think the open time for the date string "2021-03-12" would be 2021-03-11 at 16:00 CET and the close time would be 2021-03-12 at 16:00 CET (or 14:15 if you prefer the compute time).

Thoughts?

**Review by mrice32 — APPROVED** [review]
LGTM!

File: packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js (hunk 7 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/FallBackPriceFeed.js]: Yeah. Having the notion of "meta pricefeeds" such as Fallback and Medianizer makes the `update`-`getPrice` pattern inefficient, whereas a more robust `async getPrice` method would be better for the meta pricefeeds. This way the meta pricefeeds don't have to blindly update all constituent price feeds

File: packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js (hunk 11 above)
Comment (nicholaspai) [reply packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js]: Great point, check out [this commit](https://github.com/UMAprotocol/protocol/pull/2718/commits/c7bf6695d06d7add6ff41fd3648cefa23e41831a)

**Review by mrice32 — APPROVED** [review]
LGTM!

File: packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js (hunk 11 above)
Comment (mrice32) [reply packages/financial-templates-lib/src/price-feed/ForexDailyPriceFeed.js]: This is really really elegant. +1
